src/Interpreters/HashJoin.cpp

```diff
@@ -827,6 +830,25 @@ bool HashJoin::addBlockToJoin(const Block & source_block_, bool check_limits)
     if (shrink_blocks)
         block_to_save = block_to_save.shrinkToFit();
 
+    size_t max_bytes_in_join = table_join->sizeLimits().max_bytes;
+    size_t max_rows_in_join = table_join->sizeLimits().max_rows;
+    if (!table_join->sizeLimits().hasLimits())
+    {
+        max_bytes_in_join = table_join->defaultMaxBytes();
+    }
+
+    if (kind == JoinKind::Cross && tmp_data
+        && (tmp_stream || (max_bytes_in_join && getTotalByteCount() + block_to_save.allocatedBytes() >= max_bytes_in_join)
```

Comment:
It's also good to add test for this specific scenario. Check out examples in 
https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless
https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/1_stateful (these tests has an access to preloaded `hits` and `visits` tables)

for instance:

https://github.com/ClickHouse/ClickHouse/blob/26846bd238813c56ad36654ce6dac4b57e17e0e1/tests/queries/0_stateless/01010_pmj_right_table_memory_limits.sql